### PR TITLE
Feature/schema versions

### LINKED ISSUE #2670: Add schema versions and dbt versions to json output
### Describe the feature
Add schema versions to our hologram output. I imagine we'd version them with the [$schema keyword](https://json-schema.org/understanding-json-schema/reference/schema.html), but I'm open to using an explicit `schema-version` field instead.

I've separated this from #2671 as I think using `$schema` means we should host the schemas somewhere and that requires a bit of infrastructure-type work. Or if we don't want to host the schemas, at least requires some care when it comes to generating the URLs.

### Describe alternatives you've considered
No versions! Anarchy!

### Who will this benefit?
Consumers of dbt's json output.

### PR DESCRIPTION
**Pull request by beckjake**
resolves #2670 

### Description
Add schema versions and dbt versions dbt's artifact output.

Long-term, these belong in the `metadata` field described in #2761.

I think I would like to get some automated mechanism for this into either hologram or the class, rather than this method of manually inserting things.

We can definitely change the name to something better, I don't love `dbt_schema_version`.

### Checklist
 - [X] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
 - [X] I have run this code in development and it appears to resolve the stated issue
 - [X] This PR includes tests, or tests are not required/relevant for this PR
 - [X] I have updated the `CHANGELOG.md` and added information about my change to the "dbt next" section.

### CODE REVIEW

**Review by gshank — APPROVED** [review]

File: core/dbt/contracts/state.py [hunk 1]
```diff
@@ -10,4 +11,8 @@ def __init__(self, path: Path):
 
         manifest_path = self.path / 'manifest.json'
         if manifest_path.exists() and manifest_path.is_file():
-            self.manifest = WritableManifest.read(str(manifest_path))
+            try:
+                self.manifest = WritableManifest.read(str(manifest_path))
+            except IncompatibleSchemaException as exc:
```
Comment (jtcohen6) [inline core/dbt/contracts/state.py:16]: 👍

File: core/dbt/exceptions.py [hunk 2]
```diff
@@ -257,6 +257,33 @@ def __reduce__(self):
...
+        self.msg = self.get_message()
+
+    def get_message(self) -> str:
+        found_str = 'nothing'
+        if self.found is not None:
+            found_str = f'"{self.found}"'
+
+        msg = (
+            f'Expected a schema version of "{self.expected}" in '
+            f'{self.filename}, but found {found_str}'
```
Comment (jtcohen6) [inline core/dbt/exceptions.py:279]: What do you think of offering a hint to users who might be initially confused about what `schema version` means?
```
  Expected a schema version of "https://schemas.getdbt.com/dbt/manifest/v1.json" in
.state/manifest.json, but found "https://schemas.getdbt.com/dbt/manifest/v2.json".
Are you running with a different version of dbt?
```

**Review by jtcohen6 — COMMENTED** [review]
I don't mind `dbt_schema_version`. That said, if you did want to change it, I'd support something even more explicit, like `dbt_artifact_schema_version`.

I tested locally and saw `dbt_version` and `dbt_schema_version` propagated in every artifact _except_ `sources.json`. Is that known, and to come in a future PR?

For my own edification: would our plan be to keep schema versions consistent across all artifacts, even if we only need to change one or two? Or is there a possibility of (e.g.) a v4 manifest, v3 run results, v2 catalog?

**Comment by beckjake** [comment]
>propagated in every artifact except `sources.json`. Is that known, and to come in a future PR?

This will be fixed in the metadata PR. I didn't realize it was relevant here, but the freshness output has its own special schema that _isn't_ the freshness results, which is used solely for writing, and that doesn't get the schema version. I'm going to coerce it into the standard-ish format as part of that.

>would our plan be to keep schema versions consistent across all artifacts, even if we only need to change one or two? Or is there a possibility of (e.g.) a v4 manifest, v3 run results, v2 catalog?

I hadn't planned on keeping them consistent, though I can see the value in it. I think mixed versions are pretty reasonable things to have. If we change the catalog and your code only cares about the manifest, why should you get an error? This is especially relevant for `defer`, I think.

**Comment by jtcohen6** [comment]
> If we change the catalog and your code only cares about the manifest, why should you get an error? This is especially relevant for `defer`, I think.

Fair!

**Review by jtcohen6 — APPROVED** [review]

**Pull request merged**